File: 20-actors/niyaku/methods/stow_plan.py
"""stow_plan — container stowage slotting + discharge/load sequencing.

A container ship cell is addressed by **bay / row / tier** (ISO stowage
coordinates). Automated handling needs two plan products:

  1. a **stow plan** — which slot each box occupies, respecting
     weight-on-top, hazmat segregation, reefer-plug slots, and port-rotation
     (a box for an earlier discharge port must not be buried under a later one);
  2. a **work sequence** — the order the STS crane discharges/loads boxes so it
     never has to re-handle (lift a box only to put it back to reach another).

stdlib-only · pywasm-ready. Pure planning compute; emits no outward action.

Per ADR-2606074000 (niyaku R0).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple


@dataclass(frozen=True)
class Slot:
    """A cell address. tier increases upward (00 = bottom)."""
    bay: int
    row: int
    tier: int

    def key(self) -> Tuple[int, int, int]:
        return (self.bay, self.row, self.tier)


@dataclass
class Container:
    box_id: str
    weight_t: float
    discharge_port: str          # port code where this box leaves the ship
    reefer: bool = False         # needs a powered (reefer) slot
    hazmat: Optional[str] = None # IMDG class, e.g. "3" flammable; None if not


class StowError(ValueError):
    """Raised on an infeasible stow request (weight/segregation/capacity)."""


@dataclass
class StowagePlan:
    assignments: Dict[str, Slot] = field(default_factory=dict)
    # discharge order: list of port codes in the vessel's rotation
    rotation: List[str] = field(default_factory=list)

    def slot_of(self, box_id: str) -> Slot:
        return self.assignments[box_id]


def _stack_columns(bays: int, rows: int, tiers: int) -> List[Tuple[int, int]]:
    return [(b, r) for b in range(bays) for r in range(rows)]


def build_stow_plan(
    containers: List[Container],
    rotation: List[str],
    bays: int,
    rows: int,
    tiers: int,
    reefer_rows: Optional[List[int]] = None,
) -> StowagePlan:
    """Assign every container a slot under the core stowage constraints.

    Constraints enforced:
      * capacity — at most ``bays*rows*tiers`` boxes;
      * port rotation — a box discharged at an EARLIER port is stacked ABOVE one
        discharged later in the same column (so it is lifted first, no rehandle);
      * weight-on-top — no heavier box rests on a lighter one in a column;
      * reefer — reefer boxes only in ``reefer_rows`` (if given);
      * hazmat — two different IMDG classes never share a column (segregation).

    Raises ``StowError`` if it cannot place all boxes.
    """
    if not rotation:
        raise StowError("rotation must list at least one discharge port")
    reefer_rows = reefer_rows if reefer_rows is not None else list(range(rows))
    rot_index = {p: i for i, p in enumerate(rotation)}
    for c in containers:
        if c.discharge_port not in rot_index:
            raise StowError(f"{c.box_id}: discharge_port {c.discharge_port} not in rotation")

    # Sort: latest-discharge first (goes to the BOTTOM of a column), then
    # heaviest first (heavy on the bottom). Stable so equal keys keep input order.
    order = sorted(
        containers,
        key=lambda c: (-rot_index[c.discharge_port], -c.weight_t),
    )

    columns = _stack_columns(bays, rows, tiers)
    # per-column running state
    col_height: Dict[Tuple[int, int], int] = {col: 0 for col in columns}
    col_hazmat: Dict[Tuple[int, int], Optional[str]] = {col: None for col in columns}
    col_top_weight: Dict[Tuple[int, int], float] = {col: float("inf") for col in columns}
    col_top_port: Dict[Tuple[int, int], int] = {col: -1 for col in columns}

    plan = StowagePlan(rotation=list(rotation))
    for c in order:
        placed = False
        for (b, r) in columns:
            col = (b, r)
            if col_height[col] >= tiers:
                continue
            if c.reefer and r not in reefer_rows:
                continue
            if c.hazmat is not None and col_hazmat[col] not in (None, c.hazmat):
                continue
            # weight-on-top: new box must be <= the box currently on top
            if c.weight_t > col_top_weight[col]:
                continue
            # rotation: box on top of a column must discharge no later than the
            # one below (earlier port = smaller index = lifted first / higher).
            if col_top_port[col] >= 0 and rot_index[c.discharge_port] > col_top_port[col]:
                continue
            tier = col_height[col]
            plan.assignments[c.box_id] = Slot(b, r, tier)
            col_height[col] = tier + 1
            col_top_weight[col] = c.weight_t
            col_top_port[col] = rot_index[c.discharge_port]
            if c.hazmat is not None:
                col_hazmat[col] = c.hazmat
            placed = True
            break
        if not placed:
            raise StowError(f"no feasible slot for {c.box_id}")
    return plan


def discharge_sequence(plan: StowagePlan, port: str) -> List[str]:
    """Order to discharge all boxes for ``port``: top tier first, per column.

    Guarantees no re-handle: within each column boxes are lifted top→bottom, and
    the stow plan already placed earlier-discharge boxes higher.
    """
    by_col: Dict[Tuple[int, int], List[Tuple[int, str]]] = {}
    # need discharge port per box → recover from rotation isn't stored per box,
    # so callers pass the subset; here we sequence whatever is in `plan`.
    for box_id, slot in plan.assignments.items():
        by_col.setdefault((slot.bay, slot.row), []).append((slot.tier, box_id))
    seq: List[str] = []
    # higher tier first within a column; columns in (bay,row) order
    for col in sorted(by_col):
        for _tier, box_id in sorted(by_col[col], key=lambda t: -t[0]):
            seq.append(box_id)
    return seq


def validate_no_rehandle(plan: StowagePlan, rotation_index: Dict[str, int],
                         box_port: Dict[str, str]) -> bool:
    """True iff no column has a later-discharge box stacked above an earlier one."""
    by_col: Dict[Tuple[int, int], List[Tuple[int, str]]] = {}
    for box_id, slot in plan.assignments.items():
        by_col.setdefault((slot.bay, slot.row), []).append((slot.tier, box_id))
    for col, items in by_col.items():
        items.sort(key=lambda t: t[0])  # bottom→top
        prev = None
        for _tier, box_id in items:
            p = rotation_index[box_port[box_id]]
            if prev is not None and p > prev:
                return False  # box above discharges later than the one below
            prev = p
    return True

File: 20-actors/niyaku/methods/stow_plan.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/stow_plan.py.
(ns niyaku.methods.stow-plan
  "stow_plan — container stowage slotting + discharge/load sequencing (ADR-2606082000, niyaku R0).

  A container ship cell is addressed by bay / row / tier (ISO stowage coords). Automated
  handling needs a STOW PLAN (which slot each box occupies — respecting weight-on-top, IMDG
  hazmat segregation, reefer-plug rows, and port-rotation so an earlier-discharge box is never
  buried under a later one) and a WORK SEQUENCE (discharge order with no re-handle).

  Pure planning compute; emits no outward action. G9: an infeasible request RAISES (stow-error),
  never silently relaxes a safety constraint. reefer rows = the cold-chain (food/seafood) plug
  slots — the food-logistics tie-in.

  Run:  bb --classpath 20-actors 20-actors/niyaku/methods/stow_plan.clj"
  (:require [clojure.string :as str]))

;; a slot {:bay :row :tier} (tier increases upward, 0 = bottom); a container
;; {:box-id :weight-t :discharge-port :reefer :hazmat}.
(defn slot [bay row tier] {:bay bay :row row :tier tier})
(defn slot-key [s] [(:bay s) (:row s) (:tier s)])

(defn container [box-id weight-t discharge-port & {:keys [reefer hazmat] :or {reefer false hazmat nil}}]
  {:box-id box-id :weight-t (double weight-t) :discharge-port discharge-port
   :reefer reefer :hazmat hazmat})

(defn stow-error [msg] (throw (ex-info msg {:type :stow-error})))
(defn stow-error? [e]
  (and (instance? clojure.lang.ExceptionInfo e) (= :stow-error (:type (ex-data e)))))

(defn slot-of [plan box-id] (get-in plan [:assignments box-id]))

(defn build-stow-plan
  "Assign every container a slot under capacity / port-rotation / weight-on-top / reefer-row /
  IMDG-segregation constraints. Raises (stow-error) if it cannot place all boxes."
  [containers rotation bays rows tiers & {:keys [reefer-rows] :or {reefer-rows ::all}}]
  (when (empty? rotation) (stow-error "rotation must list at least one discharge port"))
  (let [reefer-rows (set (if (= reefer-rows ::all) (range rows) reefer-rows))
        rot-index (into {} (map-indexed (fn [i p] [p i]) rotation))]
    (doseq [c containers]
      (when-not (contains? rot-index (:discharge-port c))
        (stow-error (str (:box-id c) ": discharge_port " (:discharge-port c) " not in rotation"))))
    ;; latest-discharge first (to the BOTTOM), then heaviest first; stable for ties.
    (let [order (sort-by (fn [c] [(- (rot-index (:discharge-port c))) (- (:weight-t c))]) containers)
          columns (vec (for [b (range bays) r (range rows)] [b r]))
          col-height (atom (zipmap columns (repeat 0)))
          col-hazmat (atom (zipmap columns (repeat nil)))
          col-top-weight (atom (zipmap columns (repeat Double/POSITIVE_INFINITY)))
          col-top-port (atom (zipmap columns (repeat -1)))
          assignments (atom {})]
      (doseq [c order]
        (let [placed (atom false)]
          (doseq [[b r :as col] columns :while (not @placed)]
            (when (and (< (@col-height col) tiers)
                       (or (not (:reefer c)) (contains? reefer-rows r))
                       (or (nil? (:hazmat c))
                           (nil? (@col-hazmat col))
                           (= (@col-hazmat col) (:hazmat c)))
                       (<= (:weight-t c) (@col-top-weight col))
                       (or (< (@col-top-port col) 0)
                           (<= (rot-index (:discharge-port c)) (@col-top-port col))))
              (let [tier (@col-height col)]
                (swap! assignments assoc (:box-id c) (slot b r tier))
                (swap! col-height assoc col (inc tier))
                (swap! col-top-weight assoc col (:weight-t c))
                (swap! col-top-port assoc col (rot-index (:discharge-port c)))
                (when (:hazmat c) (swap! col-hazmat assoc col (:hazmat c)))
                (reset! placed true))))
          (when-not @placed (stow-error (str "no feasible slot for " (:box-id c))))))
      {:assignments @assignments :rotation (vec rotation)})))

(defn- columns-of [plan]
  (group-by (fn [[_ s]] [(:bay s) (:row s)]) (seq (:assignments plan))))

(defn discharge-sequence
  "Order to discharge boxes: top tier first within each column; columns in (bay,row) order.
  No re-handle (the stow plan already placed earlier-discharge boxes higher)."
  [plan _port]
  (let [by-col (columns-of plan)]
    (vec (mapcat (fn [col]
                   (map first (sort-by (fn [[_ s]] (- (:tier s))) (get by-col col))))
                 (sort (keys by-col))))))

(defn validate-no-rehandle
  "True iff no column has a later-discharge box stacked above an earlier one."
  [plan rotation-index box-port]
  (every?
   (fn [items]
     (loop [prev nil
            xs (sort-by (fn [[_ s]] (:tier s)) items)]   ; bottom→top
       (if (empty? xs)
         true
         (let [[box-id _] (first xs)
               p (rotation-index (box-port box-id))]
           (if (and prev (> p prev))
             false
             (recur p (rest xs)))))))
   (vals (columns-of plan))))

(defn main [& _]
  (let [rotation ["SHA" "SIN" "ROT"]
        boxes [(container "A" 22.0 "ROT") (container "B" 18.0 "SIN") (container "C" 14.0 "SHA")]
        plan (build-stow-plan boxes rotation 1 1 3)]
    (println "niyaku stow plan (bay/row/tier):")
    (doseq [[bid s] (sort-by key (:assignments plan))]
      (println (format "  %s → bay %d row %d tier %d" bid (:bay s) (:row s) (:tier s))))
    (println "discharge sequence (top-first):" (discharge-sequence plan "ROT"))))

(when (= *file* (System/getProperty "babashka.file"))
  (main))
